### PR TITLE
Homework #3

### PR DESCRIPTION
**Pull request by DarkAriesX**
authors: @DarkAriesX @kari0036

### CODE REVIEW

File: module3/homework/passwordCheck/validation.cpp [hunk 1]
```diff
@@ -1,2 +1,57 @@
 #include "validation.hpp"
-// TODO: Put implementations here
\ No newline at end of file
+
+#include <algorithm>
+
+namespace{
+    size_t const minimumPasswordLength = 9;
```
Comment (DamiXioR) [inline module3/homework/passwordCheck/validation.cpp:6]: i think it's unnecessery, you use 9 just one time in code

File: module3/homework/passwordCheck/validation.hpp [hunk 2]
```diff
@@ -1,2 +1,17 @@
-// TODO: I'm empty :) Put enum and function headers here.
-// Don't forget the header guard
\ No newline at end of file
+#pragma once
+
+#include <iostream>
```
Comment (DamiXioR) [inline module3/homework/passwordCheck/validation.hpp:3]: you use it only in implementation file

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#include "vectorFunctions.hpp"
+
+std::vector<std::shared_ptr<int>> generate(int count){
+    std::vector<std::shared_ptr<int>> pointers;
+    pointers.reserve(count);
+    
+    int i = 0;
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:7]: i think it's better to use for loop

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+void add10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        if(el == nullptr){
+            break;
+        }
+        *el = *el + 10;
+    }
+}
+void sub10(int * ptr){
+    if(ptr != nullptr){
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:30]: (ptr) is sufficient

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+#include <iostream>
+#include <string>
+#include <vector>
+#include <memory>
+
+std::vector<std::shared_ptr<int>> generate(int);
+void print(const std::vector<std::shared_ptr<int>>&);
+void add10(std::vector<std::shared_ptr<int>>&);
+void sub10(int *);
+void sub10(std::vector<std::shared_ptr<int>>&);
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:12]: you don't change vector so you can use const&

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+}
+void add10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        if(el == nullptr){
+            break;
+        }
+        *el = *el + 10;
+    }
+}
+void sub10(int * ptr){
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:29]: it shoulb be const

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    if(ptr != nullptr){
+        *ptr = *ptr - 10;
+    }
+}
+void sub10(std::vector<std::shared_ptr<int>>& vec){
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:34]: use const reference

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include <iostream>
+#include <string>
```
Comment (DamiXioR) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:4]: move string, vector and memory to header file

File: module3/homework/passwordCheck/validation.cpp [hunk 9]
```diff
@@ -1,2 +1,57 @@
...
+   }
+
+   if (std::all_of(password.begin(), password.end(), ::isalnum)) {
+      return ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter;
+   }
+
+   if (std::none_of(password.begin(), password.end(), ::isupper)) {
+      return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+      }
+return ErrorCode::Ok;
```
Comment (kszytko) [inline module3/homework/passwordCheck/validation.cpp:47]: Missed indent and enter before this line.

File: module3/homework/passwordCheck/validation.cpp [hunk 10]
```diff
@@ -1,2 +1,57 @@
...
+}
+
+ErrorCode checkPassword(const std::string& first, const std::string& secound){
```
Comment (kszytko) [inline module3/homework/passwordCheck/validation.cpp:50]: *second

File: module3/homework/passwordCheck/validation.cpp [hunk 11]
```diff
@@ -1,2 +1,57 @@
...
+   if(!doesPasswordsMatch(first, secound)) {
+      return ErrorCode::PasswordsDoesNotMatch;
+   }
+
+return checkPasswordRules(first);
```
Comment (kszytko) [inline module3/homework/passwordCheck/validation.cpp:55]: Same indent missing

File: module3/homework/passwordCheck/validation.cpp [hunk 12]
```diff
@@ -1,2 +1,57 @@
 #include "validation.hpp"
-// TODO: Put implementations here
\ No newline at end of file
+
+#include <algorithm>
+
+namespace{
+    size_t const minimumPasswordLength = 9; 
+};
```
Comment (kszytko) [inline module3/homework/passwordCheck/validation.cpp:7]: Const you can place in hpp file.

File: module3/homework/vectorOfSharedPointers/main.cpp [hunk 13]
```diff
@@ -1,5 +1,3 @@
-#include <iostream>
-#include <string>
```
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/main.cpp:2]: Dont modify main.cpp if not needed.

File: module3/homework/passwordCheck/main.cpp [hunk 14]
```diff
@@ -1,5 +1,3 @@
-#include <iostream>
-#include <string>
```
Comment (kszytko) [inline module3/homework/passwordCheck/main.cpp:2]: Dont modify main.cpp if not needed.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    for(auto const& el: vec){
+        std::cout << *el << "\n";
+    }
+}
+void add10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        if(el == nullptr){
+            break;
+        }
+        *el = *el + 10;
```
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:26]:
```suggestion
        if(el){
            *el += 10;
        }
```
Instead of break, put increment into if statement.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp (hunk 4 above)
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:30]:
```suggestion
    if(!ptr){
```

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+}
+void sub10(int * ptr){
+    if(ptr != nullptr){
+        *ptr = *ptr - 10;
+    }
+}
+void sub10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        int * ptr = el.get();
+        sub10(ptr);
```
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:37]: Instead of creating new ptr, pass adress into function.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+#include <vector>
+#include <memory>
```
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:6]: Sort includes.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+
+std::vector<std::shared_ptr<int>> generate(int);
+void print(const std::vector<std::shared_ptr<int>>&);
+void add10(std::vector<std::shared_ptr<int>>&);
+void sub10(int *);
```
Comment (kszytko) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:11]: It's not a constant pointer

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp (hunk 3 above)
Comment (kszytko) [reply module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:7]: While is good as well

Comment (DamiXioR) [reply module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:7]: but you have to create redundant variable

File: module3/homework/passwordCheck/validation.cpp (hunk 1 above)
Comment (kyrtaq) [inline module3/homework/passwordCheck/validation.cpp:6]: i think it`s good to use consts instead of magic numbers so good job!

File: module3/presentation_references.md [hunk 19]
```diff
@@ -19,7 +19,7 @@ int value = 5;
 int & number = value;
 ```
 
-Powyższy zapis oznacza zmienną `num` typu `int&`, czyli referencję na typ `int`.
+Powyższy zapis oznacza zmienną `number` typu `int&`, czyli referencję na typ `int`.
 <!-- .element: class="fragment fade-in" -->
 
 Nie ma znaczenia, czy referencję dokleimy do typu, czy nazwy zmiennej, ale referencja jest oddzielnym typem, więc sugerujemy nie doklejać jej do nazwy zmiennej.
```
Comment (kyrtaq) [inline module3/presentation_references.md:25]: if you want to change things in presentation then make another PR changing it

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+void print(const std::vector<std::shared_ptr<int>>& vec){
+    for(auto const& el: vec){
+        std::cout << *el << "\n";
+    }
+}
+void add10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        if(el == nullptr){
+            break;
+        }
```
Comment (Vicx95) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:25]:
```suggestion
void add10(const std::vector<std::shared_ptr<int>>& vec){
    for(const auto& el: vec){
        if(el == nullptr){
            break;
        }
```

File: module3/homework/passwordCheck/validation.cpp (hunk 1 above)
Comment (kszytko) [reply module3/homework/passwordCheck/validation.cpp:6]: It is good practise to use const values instead magic numbers in code. And better to learn it now.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp (hunk 3 above)
Comment (kszytko) [reply module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:7]: In for loop you create redundant as well. Ok it is local variable but compiler need to alocate memory like in while loop.

File: module3/homework/passwordCheck/validation.cpp (hunk 1 above)
Comment (nauka-programowania-MA) [reply module3/homework/passwordCheck/validation.cpp:6]: Yes its good practice, but when you use this value only in one function, please move this to this function.

File: module3/homework/passwordCheck/validation.cpp (hunk 9 above)
Comment (nauka-programowania-MA) [reply module3/homework/passwordCheck/validation.cpp:47]: +1

File: module3/homework/passwordCheck/validation.hpp (hunk 2 above)
Comment (nauka-programowania-MA) [inline module3/homework/passwordCheck/validation.hpp:3]: move to .cpp file

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp (hunk 3 above)
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:7]: When you need to declare `i` to while loop, it's mean you want use for loop :)

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    return pointers;
+}
+void print(const std::vector<std::shared_ptr<int>>& vec){
+    for(auto const& el: vec){
+        std::cout << *el << "\n";
+    }
+}
+void add10(std::vector<std::shared_ptr<int>>& vec){
```
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:21]: const vector. You don't modify vector, you modify values pointed by `shared`ptr`.

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    }
+}
+void sub10(int * ptr){
+    if(ptr != nullptr){
+        *ptr = *ptr - 10;
+    }
+}
+void sub10(std::vector<std::shared_ptr<int>>& vec){
+    for(auto& el: vec){
+        int * ptr = el.get();
```
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:36]: Don't need temporary value, just pass this to function

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp [hunk 23]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include <iostream>
```
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:3]: remove

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp (hunk 18 above)
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:11]: use const

File: module3/homework/vectorOfSharedPointers/vectorFunctions.hpp (hunk 5 above)
Comment (nauka-programowania-MA) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.hpp:12]: const

File: module3/presentation_references.md (hunk 19 above)
Comment (nauka-programowania-MA) [reply module3/presentation_references.md:25]: +1

File: module3/homework/vectorOfSharedPointers/vectorFunctions.cpp [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        sub10(ptr);
+    }
+}
```
Comment (kyrtaq) [inline module3/homework/vectorOfSharedPointers/vectorFunctions.cpp:39]: new line

File: module3/homework/passwordCheck/validation.cpp [hunk 25]
```diff
@@ -1,2 +1,60 @@
 #include "validation.hpp"
-// TODO: Put implementations here
\ No newline at end of file
+
+#include <iostream>
+#include <algorithm>
+#include <string>
```
Comment (WBigaj) [inline module3/homework/passwordCheck/validation.cpp:5]: You have included string in the header.You don't need to include this one more time.

File: module3/homework/passwordCheck/validation.cpp [hunk 26]
```diff
@@ -1,2 +1,60 @@
...
+      return ErrorCode::PasswordNeedsAtLeastOneNumber;
+   }
+
+   if (std::all_of(password.begin(), password.end(), ::isalnum)) {
+      return ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter;
+   }
+
+   if (std::none_of(password.begin(), password.end(), ::isupper)) {
+      return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+      }
```
Comment (WBigaj) [inline module3/homework/passwordCheck/validation.cpp:49]: Minor formatting issue with the bracket.

**Comment by ziobron** [comment]
+7 XP Podstawy#3 - passwordCheck (1995)
+4 XP Podstawy#3 – passwordCheck bonus requirements (1996)
+5 XP Podstawy#3 – vectorOfSharedPointers (1998)
+4 XP for delivering on time (2 x 1119)
+3 XP for cooperation (1636)

{
"requirement_ids": [1119, 1995, 1996, 1119, 1998, 1636],
"points": [2, 7, 4, 2, 5, 3],
"users": ["DarkAriesX", "kari0036"]
}
